### ISSUE TITLE
Network Action not being executed although App is Online

### ISSUE THREAD

**Issue by greatwillow** [redux-offline/redux-offline#132]
I'm newer to redux-offline, and so may be missing something, though I have implemented a standard request and network actions are not being executed, despite being online.  The following is my code alongside a screenshot of the debugging console showing online to be 'true' while retry efforts are made.  Any guidance would be appreciated!

```export const postNewUserSignup = userEmail => ({
  type: actionTypes.POST_NEW_USER_SIGNUP,
  payload: { userEmail },
  meta: {
    offline: {
      // the network action to execute:
      effect: {
        url: "https://damp-tor-16286.herokuapp.com/users/create-user",
        method: "POST",
        body: { userEmail }
      },
      // action to dispatch when effect succeeds:
      commit: {
        type: actionTypes.NEW_USER_SIGNUP_SUCEEDED,
        meta: { userEmail }
      },
      // action to dispatch if network action fails permanently:
      rollback: {
        type: actionTypes.NEW_USER_SIGNUP_FAILED,
        meta: { userEmail }
      }
    }
  }
});
```
![screen shot 2017-10-26 at 6 06 11 pm](https://user-images.githubusercontent.com/15651001/32079656-8aa4bc38-ba79-11e7-8091-b139d1d4ad69.png)

**Comment by sorodrigo** [comment]
Try stringifying the body of the effect.

**Comment by jansel369** [comment]
This works for me by supplying a commit and rollback meta to the action

**Comment by greatwillow** [comment]
Thanks for the feedback!... though not yet having any luck unfortunately.  I tried stringifying the body of the effect, though this did not change things.  @jansel369 , I'm a little unsure about what you mean by 'supplying a commit and rollback meta to the action' since I believe this is what I've already done in the above code.  Would you be able to elaborate upon this?  Perhaps I am missing something?  Thanks!

**Comment by sorodrigo** [comment]
Check this out https://github.com/redux-offline/redux-offline/issues/16

**Comment by greatwillow** [comment]
Ok!... So checked this out, and have tried multiple solutions related to what is presented here with regards to the middleware configuration, though still not having much luck.  Actually, I've tried now without any other middleware and I'm still getting the same result, so this makes me wonder if it is a middleware issue or something else.  Any other suggestions?  Thanks!

**Comment by sorodrigo** [comment]
The other thing I can think of is that you have a javascript error in the commit reducer. This is an existing issue that has been fixed in https://github.com/redux-offline/redux-offline/pull/3

**Comment by greatwillow** [comment]
Ok!  Checked this out, though quite sure the reducer code is not the issue here.  I've tried different variations now including a variation of the one from the docs with no luck yet.  At the moment I am using React-Native 0.49, and I noticed in their docs for the NetInfo API that both the .fetch() method has been deprecated for the getConnectionInfo() method while the 'change'  event has been deprecated for the 'connectionChange' event.  I suspected earlier that these updates may have resulted in this circumstance, though upon modifying the code in detectNetwork.native.js and even replacing the NetInfo.js file with one from a previous version of RN, I didn't make progress.  Could this be related here?  Regardless, as an aside, it seems that the detectNetwork.native.js file may need to be modified in the repository to account for these NetInfo updates in the latest versions of RN.  Thanks!

**Comment by sorodrigo** [comment]
Hey! The detectNetwork deprecated notices have been handled in the parallel repo. We'll be merging them soon. I'm pretty sure this is not the cause of your problems.

Could you provide the effect's response that ends up retrying your action? The issue you're facing is caused because the effect gets a response that is not accepted by the `discard.js` so a retry is dispatched. Could you please debug the response of the effect and what is being received by the discard? That would help narrow it down.

**Comment by celso-henrique** [comment]
Having same problem here, React Native 0.50 :(

**Comment by sorodrigo** [comment]
Could you provide more details? Or a code example? We have this https://codesandbox.io/s/8xml9l1r0j you can use it to create an example

**Comment by Diwei-Chen** [comment]
Hi guys, I met the same problem after upgraded from v2.0.0.

![screen shot 2017-12-11 at 11 18 30 am](https://user-images.githubusercontent.com/26317142/33811263-35e2ea52-de65-11e7-99dc-d96d000cfc64.png)

The action in output queue doesn't even get fired while the network status is online. But after I do a hot reload(command + r on os x), the action(promise) on the queue gets sent automatically.

**Comment by Diwei-Chen** [comment]
Confirmed that same codes work with v2.0.0 but not v2.2.1.

**Comment by connercms** [comment]
I am having the same issue. All my dispatched actions immediately go to outbox even though online is true. I also tried purging my redux store and my offline outbox still persists.v2.2.1 and react native 0.49.5

**Comment by wacii** [comment]
There's not much we can do without code samples.

**Comment by Diwei-Chen** [comment]
@wacii Thanks for the quick response.

Cos some of the codes in my project may be irrelevant so I extracted the following code snippet and hopefully this could help:

```
// In my comment component, the action is dispatched by clicking send comment button

this.props.dispatch({ type: Type.SEND_COMMENT, buzzId });

// In my redux-saga config file, I have the following codes:

function* sendCommentOfflineFirst(payload) {
  const { buzzId } = payload;
  const comment = yield select(getCurrentComment, buzzId);
  yield put({
      type: Type.ADD_TEXT_COMMENT,
      payload: { comment },
      meta: {
        offline: {
          effect: EffectTypes.ADD_TEXT_COMMENT_REQUEST,
          commit: { type: EffectTypes.ADD_TEXT_COMMENT_COMMIT, payload: { comment }, meta: { comment } },
          rollback: { type: EffectTypes.ADD_TEXT_COMMENT_ROLLBACK, payload: { comment }, meta: { comment } }
        }
      }
    });
  }

export function* watchSendCommentFlow() {
  yield takeEvery(Type.SEND_COMMENT, sendCommentOfflineFirst);
}

// Here are the settings on my redux-offline config file:

const reduxOfflineConfig = {
    ...defaultConfig,
    detectNetwork: (callback) => NetInfo.isConnected.addEventListener('connectionChange', (isInternetConnected) => {
        console.log('REDUX OFFLINE detectNetwork', "isInternetConnected", isInternetConnected);
        if (isInternetConnected) {
            return callback(true);
        } else {
            return callback(false);
        }
    }),
    persistOptions: {
        storage: AsyncStorage,
        debounce: 1000,
        blacklist: ['meteorReduxReducers.ready', 'meteorReduxReducers.reactNativeMeteorOfflineRecentlyAdded', 'meteorReduxReducers.isInvokeRecentlyAdded', 'navReducer'],
    },
    persistCallback: () => {
        MeteorStore.store.loaded();
    },
    effect: (effect, action) => {
        switch (effect) {
            case EffectTypes.ADD_TEXT_COMMENT_REQUEST:
                {
                    return addComment(action.payload.comment);
                    break;
                }
            }
    },
    retry: (action, retries) => action.meta.urgent ? 100 : 1000 * (retries + 1)
};

// The functions are:

const addComment = (comment, status = null) => {
  return new Promise((resolve, reject) => {
    checkIsLogined().then(
      (value) => {
        Meteor.call('addComment', comment, comment.resourceParentID, comment.message.substring(0, 20), comment.message.substring(0, 20), comment.extraContent, null, comment._id, null, status, function(err, res) {
          if (err) {
            console.log('CommentActions', 'addComment - Internet Connected err', err);
            reject(err);
          }
          else {
            console.log('CommentActions', 'addComment - Internet Connected res', res);
            resolve(res);
          }
        });
      }
    ).catch((err) => {
      console.log('User is not logined, err', err);
    });
  });
}

const checkIsLogined = () => {
  return new Promise((resolve, reject) => {
    const isLogined = MeteorStore.store.getState().meteorReduxReducers.isLogined;
    if (isLogined) {
      resolve(true);
    }
    else {
      Accounts.onLogin((data) => {
        resolve(data);
      })
      Accounts.onLoginFailure((err) => {
        if (err) reject(err);
      })
    }
  })
}

```

I am not sure if I am using this package correctly.
Any comments or suggestions are welcome :)

**Comment by wacii** [comment]
@Diwei-Chen I'm not sure how this code produces the results you suggested. While the initial request is triggered by an offline action (`{ type: "...", meta: { offline: { effect: {} }`), the retries are triggered by a complete retry action (`{ type: 'Offline/COMPLETE_RETRY', payload: offlineAction }`). Your code would need to 'listen' for both sorts of events.

That said, I'm not super familiar with redux saga, and I know others have had some difficulties integrating the two libraries.

**Comment by sorodrigo** [comment]
Solved in #229
